CMakeLists.txt

```diff
@@ -1,12 +1,17 @@
 cmake_minimum_required(
-   VERSION 3.5
+   VERSION 3.9
 )
 
 option(BUILD_DOC
    "Build LaTeX documentation"
    OFF
 )
 
+option(BUILD_TESTS
+   "Build the tests. Requires GTest."
+   OFF
```

Comment:
Strongly in favor of using canonical CMake variables. [`BUILD_TESTING`](https://cmake.org/cmake/help/latest/command/enable_testing.html) is the variable generally used for enabling CTests.

My guess is that IPR covers ground large enough to actually warrant a large battery of tests. Projects of similar size I've worked on used a combination of CTest wrapping Google Tests which are fairly large. Take [rocPRIM](https://github.com/ROCmSoftwarePlatform/rocPRIM/blob/develop/test/rocprim/CMakeLists.txt) for instance, where every `.cpp` test is actually a GTest with 10-100 tests inside. `block_scan` is one CTest that's instantiated with a handful of types, sizes, corner cases by GTest for a decent test matrix. It's a nice setup.

Reply:
@MathiasMagnus 
> My guess is that IPR covers ground large enough to actually warrant a large battery of tests.

Agreed.  The issue isn't whether there should be tests; but sort of tests?  Just unit tests?  More integrated tests?  What is the appropriate framework?

Reply:
XPR<->IPR conversions are prime candidates for unit tests. As Xazax said, having examples and being part of CI would ensure that they never go out of date. Implementing a few typical transformations (const propagation, common sub-expression elimination, math transforms, etc.) would really help people get started.

Reply:
@MathiasMagnus - I hope we aren't talking past each other.  See the discussions on the test infrastructure for more context for my comment.

Reply:
@GabrielDosReis I hope so too. If by "appropriate framework" we're looking for some label we can attach to the test methodology, because IPR doesn't make much sense in isolation, "more integrated" tests without pulling in 3rd parties (not integration testing strictly speaking) pretty much translates to writing example code / documentation.

If "testing" were only as much as having all the exmples currently in the docs be available as standalone executables and checking that they actually compile (they don't), that'd still be better than nothing.

If by "appropriate framework" you meant how to actually design the test infrastructure, I'd unit test the hell out of IPR<->XPR conversion, as that truly is unit testable. Following that have as many usage examples as reasonably possible, all implementing some tranformation. Input and expected output dumped as XPR may be committed to the repo beside all of the exmples. What Xazax is proposing is to write the examples in a particular style, so they are as self-documenting and understandable as possible, mitigating the need for LaTeX docs which may go out-of-sync. I haven't seen codebases written that way, it may work.

I'm currently familiarizing myself with Doxygen+Sphinx hooked up to CMake and it's working well. It's more tech (maintenance), but has decent output. With Doxygen+Sphinx, it's trivial to embed the interesting parts of the examples (without boilerplate) into ReST docs and then compile to PDF, HTML, etc.